cmd/root/exercise.go

```diff
@@ -0,0 +1,301 @@
+				s += fmt.Sprintf("%s\n", cs.Render(c.Arrow))
+				continue
+			}
+
+			s += cs.Render(string(exRune))
+			continue
+		}
+
+		// There's at least a typed character at this point...
+		typedRune := rune(typed[i])
+
+		// Is it incorrect?
+		if typedRune != exRune {
+			if exRune == c.Enter {
+				s += fmt.Sprintf("%s\n", is.Render(c.Arrow))
+			} else {
+				s += is.Render(string(exRune))
+			}
+
+			continue
+		}
+
+		s += ts.Render(string(exRune))
+	}
+
+	return
+}
+
+func (m exerciseModel) addRuneToTypedText(rn rune) exerciseModel {
+	if len(m.typedText) == len(m.exercise.text) {
+		return m
+	}
+
+	idx := len(m.typedText)
+
+	// If the next character is an Enter,
+	// then add the Enter and the following whitespace to the typedText.
+	//
+	// This provides the appearance of auto-indentation while typing.
+	if rune(m.exercise.text[idx]) == c.Enter {
+		whiteSpace := []rune{}
+		for i := len(m.typedText) + 1; i < len(m.exercise.text) && util.IsWhitespace(rune(m.exercise.text[i])); i++ {
+			whiteSpace = append(whiteSpace, rune(m.exercise.text[i]))
+		}
+		m.typedText += string(rn)
+		m.typedText += string(whiteSpace)
+	} else {
+		m.typedText += string(rn)
+	}
+	return m
+}
+
+func (m exerciseModel) deleteRuneFromTypedText() exerciseModel {
+	tex := m.typedText
+	l := len(tex)
+
+	if l <= 0 {
+		m.typedText = tex
+		return m
+	}
+
+	currRn := rune(tex[l-1])
+
+	if !util.IsWhitespace(currRn) {
+		m.typedText = tex[:l-1]
+		return m
+	}
+
+	m.typedText = tex[:l-1]
+	l = len(m.typedText)
+	i := 1
+	// move index backwards until a non-whitespace rune is found
+	for ; util.IsWhitespace(rune(m.exercise.text[l-i])); i++ {
+	}
+	currRn = rune(m.exercise.text[l-i])
+	if currRn == c.Enter {
+		// remove all runes up to and including the newline rune
+		m.typedText = tex[:l-i]
+	}
+	return m
+}
+
+func (m exerciseModel) finished() bool {
+	// If the user hasn't reached the end of the exercise,
+	// then they're not done yet.
+	l := len(m.exercise.text)
+	if len(m.typedText) < l {
+		return false
+	}
+
+	// Handle the case where the user types the last character incorrectly
+	exLast := rune(m.exercise.text[l-1])
+	typedLast := rune(m.typedText[l-1])
+
+	if exLast != typedLast {
+		return false
+	}
+	return true
+}
+
+// Converts the exercise model to a Rep, in preparation for
+// inserting it into the database.
+func exerciseModelToRep(m exerciseModel) db.Rep {
+	return db.Rep{
+		Hash:   util.MD5Hash(m.exercise.text),
+		Start:  m.events[0].Ts,
+		End:    m.events[len(m.events)-1].Ts,
+		Name:   m.exercise.name,
+		Lang:   util.Lang(m.exercise.name),
+		Wpm:    wpm(m.events),
+		Raw:    wpmRaw(m.events),
+		Dur:    duration(m.events),
+		Acc:    accuracy(m.events),
+		Miss:   numMistakes(m.events),
+		Errs:   numUncorrectedErrors(m.events),
+		Events: m.events,
+	}
+}
+func (m exerciseModel) Init() tea.Cmd {
+	return nil
+}
+
+func (m exerciseModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
+	switch msg := msg.(type) {
+	case tea.KeyMsg:
```

Comment:
Can't I just switch on the tea message, or leave from the update lop if msg.(type) is not a tea.KeyMsg? This nesting is unnecessary